src/System.IO.Ports/src/System/IO/Ports/SerialStream.Unix.cs

```diff
@@ -0,0 +1,579 @@
+        {
+            if (array == null)
+                throw new ArgumentNullException(nameof(array));
+            if (offset < 0)
+                throw new ArgumentOutOfRangeException(nameof(offset), SR.ArgumentOutOfRange_NeedNonNegNumRequired);
+            if (count < 0)
+                throw new ArgumentOutOfRangeException(nameof(count), SR.ArgumentOutOfRange_NeedNonNegNumRequired);
+            if (array.Length - offset < count)
+                throw new ArgumentException(SR.Argument_InvalidOffLen);
+            if (count == 0) return 0; // return immediately if no bytes requested; no need for overhead.
+
+            Interop.Sys.PollEvents events = Interop.Sys.PollEvents.POLLNONE;
+
+            if (timeout > 0)
+            {
+                Interop.Sys.Poll(_handle, Interop.Sys.PollEvents.POLLIN | Interop.Sys.PollEvents.POLLERR, timeout,out events);
+
+                if ((events & (Interop.Sys.PollEvents.POLLERR | Interop.Sys.PollEvents.POLLNVAL)) != 0)
+                {
+                    throw new IOException();
+                }
+
+                if ( (events & Interop.Sys.PollEvents.POLLIN) == 0)
+                {
+                    throw new TimeoutException();
+                }
+            }
+
+            int numBytes;
+            fixed (byte* bufPtr = array)
+            {
+                numBytes = Interop.Sys.Read(_handle, bufPtr + offset, count);
+            }
+
+            if (numBytes == 0)
+                throw new TimeoutException();
+
+            return numBytes;
+        }
+
+        public override void Write(byte[] array, int offset, int count)
+        {
+            Write(array, offset, count, WriteTimeout);
+        }
+
+        internal unsafe void Write(byte[] array, int offset, int count, int timeout)
+        {
+            if (_inBreak)
+                throw new InvalidOperationException(SR.In_Break_State);
+            if (array == null)
+                throw new ArgumentNullException(nameof(array));
+            if (offset < 0)
+                throw new ArgumentOutOfRangeException(nameof(offset), SR.ArgumentOutOfRange_NeedPosNum);
+            if (count < 0)
+                throw new ArgumentOutOfRangeException(nameof(count), SR.ArgumentOutOfRange_NeedPosNum);
+            if (count == 0) return; // no need to expend overhead in creating asyncResult, etc.
+            if (array.Length - offset < count)
+                throw new ArgumentException(SR.Argument_InvalidOffLen);
+            Debug.Assert(timeout == SerialPort.InfiniteTimeout || timeout >= 0, "Serial Stream Write - write timeout is " + timeout);
+
+            // check for open handle, though the port is always supposed to be open
+            if (_handle == null) throw new IOException(); //InternalResources.FileNotOpen();
+
+            int numBytes = 0;
+
+            while (count > 0)
+            {
+                Interop.Sys.PollEvents events = Interop.Sys.PollEvents.POLLNONE;
```

Comment:
nit: I rather see this in the scope that it is used.

nit: the whole poll handling, when timeout is greater than 0, is also repeated in a few places.